op-node/rollup/derive/batches.go

```diff
@@ -25,15 +26,95 @@ const (
 	BatchFuture
 )
 
-func CheckBatchEspresso() BatchValidity {
-	// TODO: verify batch constraints
-	panic("Unimplemented")
+func CheckBatchEspresso(cfg *rollup.Config, log log.Logger, l1Blocks []eth.L1BlockRef, l2SafeHead eth.L2BlockRef, batch *BatchWithL1InclusionBlock, hotshot HotShotContractProvider) BatchValidity {
+	// First, validate the headers that represent the beginning of the L2 block range of this batch.
+	jst := batch.Batch.Justification
+	if jst == nil {
+		// Espresso blocks must have a justification
+		return BatchDrop
+	}
+	startHeaders :=
+		[]espresso.Header{jst.PrevBatchLastBlock, jst.FirstBlock}
+	headerHeights :=
+		[]uint64{jst.FirstBlockNumber - 1, jst.FirstBlockNumber}
+	err := hotshot.verifyHeaders(startHeaders, headerHeights)
+
+	// In the case that the headers aren't available yet (perhaps the validator's L1 client is behind), return BatchFuture so that we can try again later
+	// If the headers are available but invalid, drop the batch
+	if err != nil {
+		return BatchFuture
+		// TODO drop the batch if the header is invalid instead of unavailable
+	}
+
+	// First, check for cases where it is valid to have any empty batch.
+	windowStart := l2SafeHead.Time + cfg.BlockTime
+	windowEnd := windowStart + cfg.BlockTime
+	payload := jst.Payload
+	prevL1Origin := l1Blocks[0]
+
+	// If Espresso did not produce any blocks in this window, an empty batch is valid.
+	// In this case, the L1 origin must be the same as the previous block.
+	if payload == nil && jst.FirstBlock.Timestamp >= windowEnd {
+		if uint64(batch.Batch.EpochNum) != prevL1Origin.Number {
+			return BatchDrop
+		}
+		return BatchAccept
+	}
+
+	// An empty batch can also be valid if the L1 origin is too far behind (either due to lag, or because HotShot skipped a block)
+	// In these cases, the L1 origin must increase by one
+	// TODO check lag case
+	skippedL1Block := jst.FirstBlock.L1Block.Number > jst.PrevBatchLastBlock.L1Block.Number+1
+	if payload == nil && skippedL1Block {
+		if uint64(batch.Batch.EpochNum) != prevL1Origin.Number+1 {
+			return BatchDrop
+		}
+		return BatchAccept
+	}
+
+	// At this point, there is no valid reason to have an empty payload
+	if payload == nil {
+		return BatchDrop
+	}
+
+	// Now validate the transactions in the batch
+	numBlocks := len(payload.NmtProofs)
+
+	// Validate the headers representing the end of the batch window
+	endHeaders :=
+		[]espresso.Header{payload.LastBlock, payload.NextBatchFirstBlock}
+	headerHeights =
+		[]uint64{jst.FirstBlockNumber + uint64(numBlocks), jst.FirstBlockNumber + uint64(numBlocks) + 1}
```

Comment:
These are both off by one, since `numBlocks` includes both the first block and the last block. For example if `FirstBlockNumber` is 0 and there is only one block, then the last block number is also 0 (the same block), not 1.

We can assume/require `numBlocks > 0`, since we eliminated the empty payload case above, and thus `numBlocks - 1` won't underflow.